process/smartContract/processorV2/vmInputV2.go

```diff
@@ -0,0 +1,204 @@
+package processorV2
+
+import (
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/data"
+	"github.com/ElrondNetwork/elrond-go-core/data/smartContractResult"
+	"github.com/ElrondNetwork/elrond-go-core/data/vm"
+	"github.com/ElrondNetwork/elrond-go/process"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+func (sc *scProcessor) createVMDeployInput(tx data.TransactionHandler) (*vmcommon.ContractCreateInput, []byte, error) {
+	deployData, err := sc.argsParser.ParseDeployData(string(tx.GetData()))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	vmCreateInput := &vmcommon.ContractCreateInput{}
+	vmCreateInput.ContractCode = deployData.Code
+	// when executing SC deploys we should always apply the flags
+	codeMetadata := sc.blockChainHook.ApplyFiltersOnCodeMetadata(deployData.CodeMetadata)
+	vmCreateInput.ContractCodeMetadata = codeMetadata.ToBytes()
+	vmCreateInput.VMInput = vmcommon.VMInput{}
+	err = sc.initializeVMInputFromTx(&vmCreateInput.VMInput, tx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	vmCreateInput.VMInput.Arguments = deployData.Arguments
+
+	return vmCreateInput, deployData.VMType, nil
+}
+
+func (sc *scProcessor) initializeVMInputFromTx(vmInput *vmcommon.VMInput, tx data.TransactionHandler) error {
+	var err error
+
+	vmInput.CallerAddr = tx.GetSndAddr()
+	vmInput.CallValue = new(big.Int).Set(tx.GetValue())
+	vmInput.GasPrice = tx.GetGasPrice()
+	vmInput.GasProvided, err = sc.prepareGasProvided(tx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isSmartContractResult(tx data.TransactionHandler) bool {
+	_, isScr := tx.(*smartContractResult.SmartContractResult)
+	return isScr
+}
+
+func (sc *scProcessor) prepareGasProvided(tx data.TransactionHandler) (uint64, error) {
+	if isSmartContractResult(tx) {
+		return tx.GetGasLimit(), nil
+	}
+
+	if sc.shardCoordinator.ComputeId(tx.GetSndAddr()) == core.MetachainShardId {
+		return tx.GetGasLimit(), nil
+	}
+
+	gasForTxData := sc.economicsFee.ComputeGasLimit(tx)
+	if tx.GetGasLimit() < gasForTxData {
+		return 0, process.ErrNotEnoughGas
+	}
+
+	return tx.GetGasLimit() - gasForTxData, nil
+}
+
+func (sc *scProcessor) createVMCallInput(
```

Comment:
we can have this as a separate struct in a next PR, dedicated solely on creating the vm input

Reply:
not sure what to do here